### ISSUE TITLE
Failure on Django 1.7+ with `django.db.models.fields.related.RelatedManager`

### ISSUE THREAD

**Issue by rhunwicks** [django-import-export/django-import-export#311]
We're trying to upgrade from Django 1.6 to 1.7 (and then to 1.8).

My tests fail with:

```
  File "/home/roger/.virtualenvs/fdw/local/lib/python2.7/site-packages/import_export/resources.py", line 383, in import_data
    real_dry_run)
  File "/home/roger/.virtualenvs/fdw/local/lib/python2.7/site-packages/import_export/resources.py", line 294, in get_diff
    v2 = self.export_field(field, current) if current else ""
  File "/home/roger/.virtualenvs/fdw/local/lib/python2.7/site-packages/import_export/resources.py", line 416, in export_field
    return field.export(obj)
  File "/home/roger/.virtualenvs/fdw/local/lib/python2.7/site-packages/import_export/fields.py", line 101, in export
    value = self.get_value(obj)
  File "/home/roger/.virtualenvs/fdw/local/lib/python2.7/site-packages/import_export/fields.py", line 86, in get_value
    value = value()
  File "/home/roger/.virtualenvs/fdw/local/lib/python2.7/site-packages/django/db/models/fields/related.py", line 693, in __call__
    manager = getattr(self.model, kwargs.pop('manager'))
KeyError: u'manager'
```

I can see that 0dfb4b9d8288bf47e56465f9269f96f56aea5c55 first tried to solve this using:

```
       # Manyrelatedmanagers are callable in Django >= 1.7 but we don't want
       # to call them
       if callable(value) and not value.through:
          value = value()
```

And that this was quickly upgraded by 2e2dc84aea199393549f21d312185cb88487eb4b to:

```
        if callable(value) and not hasattr(value, 'through'):
            value = value()
```

This still doesn't work for me. I have a model with a straightforward foreign key. It is still represented as a `django.db.models.fields.related.RelatedManager` but it isn't a ManyToMany and it doesn't have a `through` attribute and consequently it gets called here. That results in the following Django code being called:

```
def create_foreign_related_manager(superclass, rel_field, rel_model):
    class RelatedManager(superclass):
        def __init__(self, instance):
            super(RelatedManager, self).__init__()
            self.instance = instance
            self.core_filters = {rel_field.name: instance}
            self.model = rel_model

        def __call__(self, **kwargs):
            # We use **kwargs rather than a kwarg argument to enforce the
            # `manager='manager_name'` syntax.
            manager = getattr(self.model, kwargs.pop('manager'))
```

We aren't passing any arguments and the `kwargs` is an empty `dict` and so the `pop` fails.

If we don't want to call Managers, why don't we just test for them:

```
from django.db.models.manager import Manager


        # RelatedManager and ManyRelatedManager classes are callable in
        # Django >= 1.7 but we don't want to call them
        if callable(value) and not isinstance(value, Manager):
            value = value()
        return value
```

**Comment by bmihelac** [comment]
@rhunwicks does your proposed solution works? if so, can you send a pull request?

thanks!

**Comment by rhunwicks** [comment]
It does fix the error I was getting. I'm trying to reproduce it in the django-import-export test suite so I can provide a test as well as a fix, but I haven't managed it yet.

**Comment by bmihelac** [comment]
Hi @rhunwicks 

I noticed today when running tox that ``test_get_diff_with_callable_related_manager`` test fails with all django versions except 1.8. 

this is stacktrace:

    ======================================================================
    ERROR: test_get_diff_with_callable_related_manager (core.tests.resources_tests.ModelResourceTest)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/Users/bmihelac/dev/django/dev/django-import-export/tests/core/tests/resources_tests.py", line 205, in test_get_diff_with_callable_related_manager
        diff = resource.get_diff(author2, author)
      File "/Users/bmihelac/dev/django/dev/django-import-export/tests/../import_export/resources.py", line 293, in get_diff
        v1 = self.export_field(field, original) if original else ""
      File "/Users/bmihelac/dev/django/dev/django-import-export/tests/../import_export/resources.py", line 416, in export_field
        return field.export(obj)
      File "/Users/bmihelac/dev/django/dev/django-import-export/tests/../import_export/fields.py", line 105, in export
        return self.widget.render(value)
      File "/Users/bmihelac/dev/django/dev/django-import-export/tests/../import_export/widgets.py", line 36, in render
        return force_text(value)
      File "/Users/bmihelac/dev/django/dev/django-import-export/.tox/py27-1.7/lib/python2.7/site-packages/django/utils/encoding.py", line 85, in force_text
        s = six.text_type(s)
      File "/Users/bmihelac/dev/django/dev/django-import-export/.tox/py27-1.7/lib/python2.7/site-packages/django/db/models/manager.py", line 81, in __str__
        in opts.concrete_managers + opts.abstract_managers
    StopIteration

**Comment by rhunwicks** [comment]
I look at it. Thanks for the heads up.

**Comment by bmihelac** [comment]
I am having hard time to see what [AuthorResource](https://github.com/django-import-export/django-import-export/blob/4bee76cc4b2778257a983fc4fb59223174161e8f/tests/core/tests/resources_tests.py#L104-L114) in tests should export  for `books`. It fails for Django <=1.7 and for later versions it returns name of manager, ie `core.Book.None`. I have a feeling that it should be used with some kind of widget to get some meaningful output.
Can you help me @rhunwicks?

For now I skipped this test so travis does not complain.

**Comment by rhunwicks** [comment]
My problem wasn't that I wanted to export anything - it is that imports were failing because `Field.get_value()` was calling the Manager. I.e. in 1.7+ where Managers are callable. I assume that this is the reason that the check for `through` was added. However, that check only works for ManyToMany models. I had an import failing for a ForeignKey - it didn't match the check, so the Manager was being called, which generated an error. I wanted to change the check so that it matched ForeignKey and ManyToMany fields, and I wanted to create a test that demonstrated the error that I was having before I made the change.

Hope that helps?
